library/ssl_tls.c

```diff
@@ -1920,39 +1895,30 @@ static int ssl_decrypt_buf( mbedtls_ssl_context *ssl )
 #endif /* MBEDTLS_SSL_PROTO_SSL3 */
 #if defined(MBEDTLS_SSL_PROTO_TLS1) || defined(MBEDTLS_SSL_PROTO_TLS1_1) || \
     defined(MBEDTLS_SSL_PROTO_TLS1_2)
-        if( ssl->minor_ver > MBEDTLS_SSL_MINOR_VERSION_0 )
-        {
-            /*
-             * TLSv1+: always check the padding up to the first failure
-             * and fake check up to 256 bytes of padding
-             */
-            size_t pad_count = 0, real_count = 1;
-            size_t padding_idx = ssl->in_msglen - padlen - 1;
-
-            /*
-             * Padding is guaranteed to be incorrect if:
-             *   1. padlen >= ssl->in_msglen
-             *
-             *   2. padding_idx >= MBEDTLS_SSL_MAX_CONTENT_LEN +
-             *                     ssl->transform_in->maclen
-             *
-             * In both cases we reset padding_idx to a safe value (0) to
-             * prevent out-of-buffer reads.
-             */
-            correct &= ( ssl->in_msglen >= padlen + 1 );
-            correct &= ( padding_idx < MBEDTLS_SSL_MAX_CONTENT_LEN +
-                                       ssl->transform_in->maclen );
-
-            padding_idx *= correct;
-
-            for( i = 1; i <= 256; i++ )
+        if( transform->minor_ver > MBEDTLS_SSL_MINOR_VERSION_0 )
+        {
+            /* The padding check involves a series of up to 256
+             * consecutive memory reads at the end of the record
+             * plaintext buffer. In order to hide the length and
+             * validity of the padding, always perform exactly
+             * `min(256,plaintext_len)` reads (but take into account
+             * only the last `padlen` bytes for the padding check). */
+            volatile size_t pad_count = 0;
+            volatile size_t real_count = 0;
+            volatile unsigned char* const check = data;
```

Comment:
I don't think all these `volatile` qualifiers are necessary, but let's leave this question for later when this is rewritten anyway.

Reply:
Actually, with all these `volatile` qualifiers, IAR complains about the order of evaluation in the `padlen += ...` statement below, leading to CI failure.

What I attempted to remove through the qualification of `pad_count` and `real_count` was the possibility of a compiler omitting the `pad_count` increase as long as `real_count == 0`. However, with hindsight, this doesn't seem to be neither realistic nor timing significant, so I pushed a commit d65f48a to remove the volatile qualification of `pad_count` and `real_count`, keeping the one for the padding.

@mpg WDYT?